### PR TITLE
[RHCLOUD-33788] Add PagerDuty to backend, engine

### PR DESCRIPTION
**Pull request by jessicarod7**
## Jira issue

<https://issues.redhat.com/browse/RHCLOUD-33788>

## Description

This branch implements PagerDuty support in `backend` and `engine`, along with key components in `common`.

Highlights:
- A new `EndpointType.PAGERDUTY` (`"pagerduty"`) used for all events sent to PagerDuty.
- New `PagerDutyProperties` and DTO for configuring the connection. PagerDuty uses a common REST endpoint, so the user does not need to provide a URL.

Also, an updated format is used to pass events to `connector-pagerduty` (which has been updated to match):

```json5
{
  "payload": {
    // The results of `BaseTransformer.toJsonObject(Event)` go here
    "environment_url": "<environment.url>" // Used to set `client` and `client_url`
    "severity": "<properties.severity>" // Can be configured by user. Options: info, warning, error, critical
  },
  "url": "<properties.url>", // this is always https://events.pagerduty.com/v2/enqueue
  "authentication": {
    "type": "SECRET_TOKEN", // SECRET_TOKEN authn must be used, and no other methods are accepted
    "secretId": "<properties.secretToken>"
  },
}
```

## Compatibility

No impact at this stage.

## Testing

Currently updating test cases.

### CODE REVIEW

**Comment by gwenneg** [comment]
@jessicarod7 Could you please update the connector in this PR and include the hardcoded URL to the PagerDuty API?

**Comment by jessicarod7** [comment]
@gwenneg I can move it, currently it's in [PagerDutyProcessor](https://github.com/RedHatInsights/notifications-backend/blob/8c6fe7bde0f6edab98fdfb3f6cdabc536369943f/engine/src/main/java/com/redhat/cloud/notifications/processors/pagerduty/PagerDutyProcessor.java#L30). The reason I put it there is because stuff like [PagerDutyConnectorRoutesTest](https://github.com/RedHatInsights/notifications-backend/blob/fdb14ea7f6d34b2ef6e4735fe206a69583c14077/connector-pagerduty/src/test/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyConnectorRoutesTest.java) expect to pass a URL in the cloud event, and then match it later on. I could include a switch to use the PagerDuty URL by default unless one is explicitly provided, if that sounds good?

**Comment by jessicarod7** [comment]
Suggested change made and test cases added.

File: backend/src/test/java/com/redhat/cloud/notifications/models/mappers/v1/endpoint/EndpointMapperTest.java [hunk 1]
```diff
@@ -117,6 +121,10 @@ void testEndpointMapperToDTO() throws IOException, URISyntaxException {
         webhookProperties.setBearerAuthentication("bearer authentication");
         webhookProperties.setSecretToken("secret token");
 
+        final PagerDutyProperties pagerDutyProperties = new PagerDutyProperties();
+        pagerDutyProperties.setSeverity(PagerDutySeverity.ERROR);
+        pagerDutyProperties.setSecretToken("secret token");
+
         // Loop through the properties and generate the JSON for the DTO.
         for (final EndpointProperties properties : List.of(camelProperties, systemSubscriptionProperties, webhookProperties)) {
```
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/models/mappers/v1/endpoint/EndpointMapperTest.java:129]:
```suggestion
        for (final EndpointProperties properties : List.of(camelProperties, systemSubscriptionProperties, webhookProperties, pagerDutyProperties)) {
```

File: database/src/main/resources/db/migration/V1.103.0__RHCLOUD-33788_introduce_pagerduty_integration.sql [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+CREATE TABLE pagerduty_properties (
+    id uuid PRIMARY KEY
+        CONSTRAINT fk_pagerduty_properties_endpoint_id
+            REFERENCES endpoints(id)
+            ON DELETE CASCADE,
+    severity varchar(10) NOT NULL,
+    secret_token_id bigint
+);
+
+COMMENT ON COLUMN pagerduty_properties.secret_token_id IS 'the ID of the secret token data that is stored in sources';
```
Comment (gwenneg) [inline database/src/main/resources/db/migration/V1.103.0__RHCLOUD-33788_introduce_pagerduty_integration.sql:14]: Please add an empty line at the end of this file:
```suggestion
COMMENT ON COLUMN pagerduty_properties.secret_token_id IS 'the ID of the secret token data that is stored in sources';

```

File: database/src/main/resources/db/migration/V1.103.0__RHCLOUD-33788_introduce_pagerduty_integration.sql [hunk 3]
```diff
@@ -0,0 +1,14 @@
+--
+-- Name: pagerduty_properties; Type: TABLE; Schema: public; Owner: -
+--
+
+CREATE TABLE pagerduty_properties (
+    id uuid PRIMARY KEY
+        CONSTRAINT fk_pagerduty_properties_endpoint_id
+            REFERENCES endpoints(id)
+            ON DELETE CASCADE,
+    severity varchar(10) NOT NULL,
```
Comment (gwenneg) [inline database/src/main/resources/db/migration/V1.103.0__RHCLOUD-33788_introduce_pagerduty_integration.sql:10]: Why is it `10` here and `32` in the Java entity?

File: connector-pagerduty/src/main/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyCloudEventDataExtractor.java [hunk 4]
```diff
@@ -30,22 +30,22 @@ public void extract(Exchange exchange, JsonObject cloudEventData) throws Illegal
 
         validatePayload(cloudEventData);
 
-        exchange.setProperty(TARGET_URL, cloudEventData.getString(URL));
+        String providedUrl = cloudEventData.getString(URL);
+        if (providedUrl == null || providedUrl.isEmpty()) {
+            exchange.setProperty(TARGET_URL, PAGERDUTY_EVENT_V2_URL);
+        } else {
+            exchange.setProperty(TARGET_URL, providedUrl);
+        }
```
Comment (gwenneg) [inline connector-pagerduty/src/main/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyCloudEventDataExtractor.java:38]: IIUC, this is needed to make the tests pass. I do not like it but I also consider this a low priority, so let's improve the tests and get rid of that code in a follow-up PR.

File: database/src/main/resources/db/migration/V1.103.0__RHCLOUD-33788_introduce_pagerduty_integration.sql (hunk 3 above)
Comment (gwenneg) [reply database/src/main/resources/db/migration/V1.103.0__RHCLOUD-33788_introduce_pagerduty_integration.sql:10]: My bad, I mixed two different topics!

File: connector-pagerduty/src/main/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyCloudEventDataExtractor.java (hunk 4 above)
Comment (jessicarod7) [reply connector-pagerduty/src/main/java/com/redhat/cloud/notifications/connector/pagerduty/PagerDutyCloudEventDataExtractor.java:38]: come up with a solution, will include in future PR. removed an assertion from PagerDutyCloudEventDataExtractorTest that was unnecessary

**Review by gwenneg — APPROVED** [review]
LGTM

**Pull request merged**